test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java

```diff
@@ -0,0 +1,413 @@
+		}
+		return targetStatus;
+	}
+
+	/**
+	 * Instruct the child process to exit and grab its outputs.
+	 * @return exit status of process.
+	 */
+	protected int terminateTarget() {
+		int rc = -1;
+		if (!active) {
+			return 0;
+		}
+		active = false;
+		try {
+			targetInWriter.write(TargetManager.TARGETVM_STOP + '\n');
+			try {
+				targetInWriter.flush();
+			} catch (IOException e) {
+				/* ignore it */
+			}
+			errOutput += getAvailableInput(targetErrReader);
+			rc = proc.waitFor();
+			proc.destroy();
+		} catch (IOException e) { 
+			/* target closed the streams */
+		} catch (InterruptedException e) {
+			logger.error("terminateTarget failed", e);
+			rc = -1;
+		} finally {
+			try {
+				/* 
+				 * Get anything that may have been in the buffers. 
+				 * The streams may not have been closed, so read without blocking.
+				 */
+				errOutput += getAvailableInput(targetErrReader);
+				outOutput += getAvailableInput(targetOutReader);
+				logger.info("id="+targetId+" display name="+displayName+" exit status ="+rc);
+				logger.info("stdout:\n"+outOutput);
+				logger.info("stderr:\n"+errOutput);
+			} catch (IOException ioe) {
+				/* ignore */
+			}
+	
+		}
+		return rc;
+	}
+
+	/**
+	 * Wait for the child process to terminate and return its exit status
+	 * @return exit status
+	 * @throws InterruptedException
+	 */
+	int waitFor() throws InterruptedException {
+		int exitStatus = proc.waitFor();
+		return exitStatus;
+	}
+
+	/**
+	 * List the contents of the directory holding OpenJ9 attach API artifacts.
+	 */
+	static void listIpcDir() {
+		String tmpdir = System.getProperty("java.io.tmpdir")
+				+ File.separatorChar + DEFAULT_IPC_DIR;
+		logger.debug("DEBUG:" + tmpdir);
+		File ipcDir = new File(tmpdir);
+		if (ipcDir.exists()) {
+			listRecursive(System.getProperty("java.io.tmpdir"), ipcDir);
+		}
+	}
+
+	/**
+	 * List the contents of  directory, including subdirectories.
+	 * @param prefix
+	 * @param root
+	 */
+	static void listRecursive(String prefix, File root) {
+		String myPath = prefix + "/" + root.getName();
+		logger.debug(myPath);
+		if (root.isDirectory()) {
+			File[] children = root.listFiles();
+			if (null != children) {
+				for (File c : children) {
+					listRecursive(myPath, c);
+				}
+			}
+		}
+	}
+
+	public static void setLogging(boolean enableLogging) {
+		doLogging = enableLogging;
+	}
+
+	public boolean isActive() {
+		return active;
+	}
+
+	private void setTargetVmid(String id) {
+		targetVmid =id;
+	}
+
+	public String getTargetVmid() {
+		return targetVmid;
+	}
+
+	public BufferedWriter getTargetInWriter() {
+		return targetInWriter;
+	}
+
+	public BufferedReader getTargetOutReader() {
+		return targetOutReader;
+	}
+
+	public BufferedReader getTargetErrReader() {
+		return targetErrReader;
+	}
+
+	/**
+	 * Reads any available lines of text from a stream without blocking, 
+	 * except to find the terminating newline.
+	 * This assumes the producer emits complete lines 
```

Comment:
`ready()` returns true as soon as a `read()` operation won't block - i.e.: a single character is available.  This comment makes assumptions that aren't guaranteed by the Class libraries.

The usual way to do this is with separate threads that read from the buffered reader and append lines to some shared data structure.  

Is there something I'm missing here?

Reply:
Discussed deskside with Peter.  This isn't new code per se, its moving old code to a new location.  I'm willing to live it, especially as it's functioned for the past ~8 years.  

We should take a hard look at some of this code when the Java 8 vs 9 vs next test code is rationalized.

Reply:
For new code, I agree with the suggestion. However, as discussed, this is cloned legacy code which we have been running for ~9 years.  Rewriting it now would add risk to this changeset.

I would be happy to re-do this code (and the Java8AndUp TargetManager) if you would open an issue.